### ISSUE TITLE
out_prometheus_exporter hang on Windows Server 2012

### ISSUE THREAD

**Issue by jefferbrecht** [fluent/fluent-bit#6275]
## Bug Report

**Describe the bug**
fluent-bit hangs on startup given the following configuration, on Windows Server 2012:

```
[OUTPUT]
    Match *
    Name  prometheus_exporter
    host  0.0.0.0
    port  20202
```

fluent-bit log:

```
[2022/10/25 13:27:13] [ info] [fluent bit] version=1.9.3, commit=9eb4996b7d, pid=1384
[2022/10/25 13:27:13] [ info] [storage] version=1.2.0, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
[2022/10/25 13:27:13] [ info] [cmetrics] version=0.3.1
```

On working systems (e.g. Server 2019) we see additional lines:

```
[2022/10/25 09:30:43] [ info] [output:prometheus_exporter:prometheus_exporter.0] listening iface=0.0.0.0 tcp_port=20202
[2022/10/25 09:30:43] [ info] [sp] stream processor started
```

But on Windows Server 2012 those lines are missing.

When running as a Windows Service, we're unable to terminate fluent-bit gracefully at this point and need to force-kill the process instead.

**To Reproduce**

Reproduction rate is >90% in my testing, but not 100% because it does rarely end up working. Happy to run experiments on my end if you have trouble reproducing this on your end.

1. Create a Windows Server 2012 machine
2. Run fluent-bit with the above config

**Expected behavior**
fluent-bit shouldn't hang

**Workaround**
We spent some time debugging this and traced it down to the following:

1. The prometheus thread creates an mk_server
2. `MK_SERVER_SIGNAL_START` is sent
3. Before `MK_SERVER_SIGNAL_START` can be consumed by the prometheus thread, the mk worker thread spawned in `mk_start` pre-empts it and steals it inside `mk_server_loop_balancer`
4. Control is _then_ yielded back to the prometheus thread which waits forever for `MK_SERVER_SIGNAL_START`, which never happens because the worker thread already consumed it

On Server 2019, the worker thread yields to the prometheus thread _before_ `mk_server_loop_balancer` is called and it all works fine.

Therefore, as a workaround we can insert a yield before calling `mk_server_loop_balancer`:

[mk_server.c:L678](https://github.com/fluent/fluent-bit/blob/master/lib/monkey/mk_server/mk_server.c#L678):

```
    else {
        /* FIXME!: this old mode needs some checks on library mode */
        printf("[%d] call loop_balancer\n", GetCurrentThreadId());
        // yield
        sleep(0);
        mk_server_loop_balancer(server);
    }
```

But because we're not experts in the monkey code, we don't know if this is an appropriate fix.

**Your Environment**

* Version used: 1.9.3 (but it also fails on 1.9.8 in our testing)
* Configuration: See above
* Environment name and version (e.g. Kubernetes? What version?): Windows Server 2012 R2
* Server type and version: Windows Server 2012 R2
* Operating System and version: Windows Server 2012 R2
* Filters and plugins: out_prometheus_exporter

**Additional context**
High automated test failure rate, and recently this is starting to affect our customers on Windows Server 2012.

**Comment by jefferbrecht** [comment]
If it helps, my environment only has 1 CPU/1 thread. I can also work around the issue by starting fluent-bit (we're running it via ops-agent) with a specific CPU affinity, e.g.

```
start /AFFINITY FF /WAIT "" "C:\Program Files\Google\Cloud Operations\Ops Agent\bin\fluent-bit.exe" -c "C:\ProgramData\Google\Cloud Operations\Ops Agent\generated_configs\fluentbit\fluent_bit_main.conf" -R "C:\ProgramData\Google\Cloud Operations\Ops Agent\generated_configs\fluentbit\fluent_bit_parser.conf" --storage_path "C:\ProgramData\Google\Cloud Operations\Ops Agent\run\buffers" --log_file "C:\ProgramData\Google\Cloud Operations\Ops Agent\log\logging-module.log"
```

**Comment by agup006** [comment]
Hey @jefferbrecht can you repro this on Windows 2019? We do not maintain older versions of Windows for Fluent Bit.

**Comment by jefferbrecht** [comment]
We can only repro this as-is on Windows 2012, however by modifying an existing `sleep` in the code we can reproduce this on any platform due to a race condition that's present on all platforms. @braydonk already fixed this in `monkey` (see the above mention) and we're proposing merging that fix back into fluent-bit.

**Comment by agup006** [comment]
Got it, so the root cause is the race condition which is showing up in slower Windows systems (2012 natively and 2019 with a sleep timer)

**Comment by braydonk** [comment]
The race condition was possible on any system if the main thread was manually stalled. We only ran into it on Windows Server 2012 because a) it uses `MK_SCHEDULER_FAIR_BALANCING_MODE` where this race condition is more likely and b) we suspect there is some kind of difference in how that version of Windows server handled context switching between threads (custom pthread implementation was hard to grok so not 100% sure).  
However, forcing the main thread to slow down hard enough could force this issue to happen on any system (I reproduced it in Linux while developing).

My fix in `monkey` was merged into fluent bit mainline as part of [4088e5fb42ac73295d675f988de2301e83e0383f](https://github.com/fluent/fluent-bit/commit/4088e5fb42ac73295d675f988de2301e83e0383f). We're hoping it will be possible to also backport it to 1.9, and I'm happy to do whatever extra testing the fluent bit folks would like us to do.
